docs/DEPRECATION.md

```diff
@@ -0,0 +1,66 @@
+# Deprecations in Vector
+
+This document covers Vector's deprecation policy and process.
+
+In the course of Vector's development it can be necessary to deprecate configuration and (rarely) features to keep
+Vector maintainable and its configuration interface consistent for users. To avoid breaking compatibility abruptly, we
+follow the following deprecation policy.
+
+## Policy
+
+Vector will retain deprecated configuration or features for at least one minor version (this will transition to one
+major version when Vector hits 1.0).
+
+This means that deprecations will be eligible for removal in the next minor release after they are announced; however,
+we will typically aim to support deprecations for a longer time period depending on their maintenance burden. For
+example, a deprecation announced in `v0.16.0` would be eligible to be removed in `v0.17.0` but may be removed later in
+`v0.20.0`.
+
+Exceptions can be made for deprecations related to security issues or critical bugs. These may result in removals being
+introduced in a release without being announced in a prior release.
+
+### Examples
+
+Examples of possible deprecations in Vector:
+
+* Removal or rename of a configuration option
+* Removal or rename of a metric
+* Removal of a feature
+
+## Notice
```

Comment:
I think it would help to organize this into stages, or the life cycle of a deprecation:

1. Notice
2. Migration
3. Enforcement

What we're really missing is the last stage, enforcement. Where we cover what the intended behavior would be after the deprecations was removed. For example, should we guarantee that Vector will error at boot-time as opposed to silently failing?

What do you think?

Reply:
I took a stab at this in e51e2fef2 . I was able to break it down into three stages, just with different names for the first and last one.